Make a dedicated struct for CppHooksList, so we can forward declare it

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28287 Merge Tensor and Variable types.
* #28282 Unfriend Variable factory functions.
* **#28240 Make a dedicated struct for CppHooksList, so we can forward declare it**
* #28227 Expunge references to Edge from AutogradMeta.

Continuing the idea of moving AutogradMeta to c10, hooks_list refers to
Variable (which I don't want to move), so I also need to forward declare it.
It's implemented as a shared_ptr (so that we can reflect updates to it
everywhere), so I can also forward declare it fine.

I do wonder if I should keep going this route of forward declaring all of the
things I don't want to move.  By the final count, there's two.  But what
if people add more metadata?  (On the other hand, AutogradMeta changes very
rarely.  So probably we're fine.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18006550](https://our.internmc.facebook.com/intern/diff/D18006550)

ezyang committed:
Make a dedicated struct for CppHooksList, so we can forward declare it

Continuing the idea of moving AutogradMeta to c10, hooks_list refers to
Variable (which I don't want to move), so I also need to forward declare it.
For some reason that I can't tell, it's implemented as a shared_ptr, so
I can also forward declare it fine.

I do wonder if I should keep going this route of forward declaring all of the
things I don't want to move.  By the final count, there's two.  But what
if people add more metadata?  (On the other hand, AutogradMeta changes very
rarely.  So probably we're fine.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make a dedicated struct for CppHooksList, so we can forward declare it"


Continuing the idea of moving AutogradMeta to c10, hooks_list refers to
Variable (which I don't want to move), so I also need to forward declare it.
For some reason that I can't tell, it's implemented as a shared_ptr, so
I can also forward declare it fine.

I do wonder if I should keep going this route of forward declaring all of the
things I don't want to move.  By the final count, there's two.  But what
if people add more metadata?  (On the other hand, AutogradMeta changes very
rarely.  So probably we're fine.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/cpp_hook.cpp:
```diff
@@ -14,15 +14,15 @@ void check_single_result (Variable value, Variable result, std::string hook_name
 
-CppFunctionPreHook::CppFunctionPreHook(const std::shared_ptr<hooks_list> &hooks, int value_idx)
-: hooks_(hooks)
+CppFunctionPreHook::CppFunctionPreHook(std::shared_ptr<CppHooksList> hooks, int value_idx)
```
Question for personal curiosity: why remove the const & and add a move here?

ezyang replied:
CppFunctionPreHook always retains a copy of CppHooksList. In this situation, pass-by-value is superior to pass-by-reference, in the case that you call this constructor with a temporary. With pass-by-reference, you have to make a copy of the temporary; here, the temporary can be directly moved into the function.

albanD reviewed (approved):
LGTM

ezyang committed:
Update on "Make a dedicated struct for CppHooksList, so we can forward declare it"


Continuing the idea of moving AutogradMeta to c10, hooks_list refers to
Variable (which I don't want to move), so I also need to forward declare it.
It's implemented as a shared_ptr (so that we can reflect updates to it
everywhere), so I can also forward declare it fine.

I do wonder if I should keep going this route of forward declaring all of the
things I don't want to move.  By the final count, there's two.  But what
if people add more metadata?  (On the other hand, AutogradMeta changes very
rarely.  So probably we're fine.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18006550](https://our.internmc.facebook.com/intern/diff/D18006550)

[ghstack-poisoned]

ezyang committed:
Update on "Make a dedicated struct for CppHooksList, so we can forward declare it"


Continuing the idea of moving AutogradMeta to c10, hooks_list refers to
Variable (which I don't want to move), so I also need to forward declare it.
It's implemented as a shared_ptr (so that we can reflect updates to it
everywhere), so I can also forward declare it fine.

I do wonder if I should keep going this route of forward declaring all of the
things I don't want to move.  By the final count, there's two.  But what
if people add more metadata?  (On the other hand, AutogradMeta changes very
rarely.  So probably we're fine.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18006550](https://our.internmc.facebook.com/intern/diff/D18006550)

[ghstack-poisoned]
